### ISSUE TITLE
Add an extension to support GitHub alerts

### ISSUE THREAD

**Issue by njzjz** [executablebooks/MyST-Parser#845]
### Describe the feature you'd like to request

Last week, GitHub supported alerts as follows:

```md
> [!NOTE]
> This is a note
```

> [!NOTE]
> This is a note

See:
- https://github.blog/changelog/2023-12-14-new-markdown-extension-alerts-provide-distinctive-styling-for-significant-content/
- https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

I hope myst-parser can add an extension to provide the same feature to provide a consistent preview for GitHub and Sphinx.

### Describe the solution you'd like

Add an option to `myst_enable_extensions` to allow such syntax.

**Comment by chrisjsewell** [comment]
Heya, it would probably be best if one first implemented a plugin in https://github.com/executablebooks/mdit-py-plugins

My only word of warning here, is that 

1. They have already flip-flopped on this, changing from the initial syntax they had before November 2023 `> **Note**` to this new one, lets hope they don't change again 😒 
2. I would agree with others in the comments for this feature; https://github.com/orgs/community/discussions/16925#discussioncomment-2794978, it basically tends to break the semantics of the blockquote syntax

Its also frustrating that Github seem to have abandoned their past spec for Github Flavoured Markdown, https://github.github.com/gfm/, when implementing these new features.

**Comment by orsinium** [comment]
+1 to the feature. While the syntax introduced by Github may suck, it is quite convenient to be able to produce documentation that is rendered in about the same way in both Github and Myst-powered web pages. Currently, the [admonitions](https://myst-parser.readthedocs.io/en/latest/syntax/admonitions.html) look bad on Github:

![image](https://github.com/executablebooks/MyST-Parser/assets/9638362/c4d466c6-e8d9-4a25-9846-1a65c0f5278e)

**Comment by henryiii** [comment]
They only flip-flopped on it while developing it, and it seems to be quite stable now and documented in their help pages.

**Comment by KyleKing** [comment]
I have a parser and render implemented in `mdformat-gfm-alerts` (https://github.com/KyleKing/mdformat-gfm-alerts?tab=readme-ov-file#mdformat-gfm-alerts), which is built on my generic admonition package. The only caveat for now is that I haven't figured out how to remove the outer block quote yet

**Comment by timkpaine** [comment]
This has been working well for me: https://github.com/python-project-templates/yardang/blob/b87fed488ac95ac99997d53fcb6cf1807b5f4ab4/yardang/conf.py.j2#L156

**Comment by KyleKing** [comment]
You should be able to add ‘mdformat-gfm-alerts’ like any other plugin (e.g. with pip install, poetry add, uv add, etc.) and it will be discovered automatically by mdformat and shouldn’t interfere with mdformat-myst 

Doing a find and replace of GFM Admonitions is also a valid approach!

**Comment by KyleKing** [comment]
Actually, a correction, for Sphinx and when rendering in general, I would probably recommend @timkpaine's approach because while `mdformat-gfm-alerts` has minimal rendering, it doesn't properly handle replacing the blockquotes. For mdformat-myst integration to format the raw markdown, I would recommend `mdformat-gfm-alerts` as described above

**Comment by ihrke** [comment]
Thanks for the advice! I have some trouble implementing @timkpaine 's advice. In my case, I use `..include ../README.md` with the myst-parser from `index.rst` to include the Github MD-file. However, it seems that the sphinx `source-read` event is not triggered with the contents of this file (it is only triggered by `index.rst` with the contents of that file). Any ideas how to fix this?

**Comment by s-weigand** [comment]
@ihrke I had the same issue with include. The solution is to connect to the `include-read` event instead of the `source-read` event.
Note that that you also need to adjust the function signature since the [`include-read` expects a different callback](https://www.sphinx-doc.org/en/master/extdev/event_callbacks.html#event-include-read)

```diff
- def run_convert_github_admonitions_to_rst(app, filename, lines):
+ def run_convert_github_admonitions_to_rst(app, relative_path, parent_docname, lines):
```

Just for completeness there is a nice [flow diagram explaining the event flow](https://www.sphinx-doc.org/en/master/extdev/event_callbacks.html#core-events-overview).

**Comment by cmarqu** [comment]
> Hmm, I tried both `source-read` and `include-read`, but neither seems to work for me. 

Never used it personally but I noticed that `include-read` was only introduced in Sphinx 7.2.5, is your version new enough?

**Comment by cmarqu** [comment]
I think you would see that if you run with -vvv.

**Comment by mara004** [comment]
I checked the `-vvv` log. Even after converting the file from Markdown to RST, there is only one occurrence of `include-read`, and that's when the callback is registered. I'm using `sphinx-build 8.2.3`.

**Comment by mara004** [comment]
Thanks for the pointer!

For the record, here's my derivation of the template function:
```python
def convert_gh_admonitions(app, relative_path, parent_docname, contents):
    # loop through content lines, replace github admonitions
    for i, orig_content in enumerate(contents):
        orig_line_splits = orig_content.split("\n")
        replacing = False
        for j, line in enumerate(orig_line_splits):
            # look for admonition key
            line_roi = line.lstrip()
            for admonition_key in _GITHUB_ADMONITIONS:
                if line_roi.startswith(admonition_key):
                    line = line.replace(admonition_key, "```{" + _GITHUB_ADMONITIONS[admonition_key] + "}")
                    # start replacing quotes in subsequent lines
                    replacing = True
                    break
            else:  # no break
                if not replacing:
                    continue
                # remove GH directive to match MyST directive
                # since we are replacing on the original line, this will preserve the right indent, if any
                if line_roi.startswith("> "):
                    line = line.replace("> ", "", 1)
                elif line_roi.rstrip() == ">":
                    line = line.replace(">", "", 1)
                else:
                    # missing "> ", so stop replacing and terminate directive
                    line = f"```\n{line}"
                    replacing = False
            # swap line back in splits
            orig_line_splits[j] = line
        # swap line back in original
        contents[i] = "\n".join(orig_line_splits)
```

A possible improvement might be handling nested directives. But even GitHub itself doesn't seem to support them yet ;)

**Comment by mara004** [comment]
Oh, the `:::` actually had a good reason: If there is a code block in the note itself, the enclosing backtickets break. So I'll change back to that.

**Comment by kdeldycke** [comment]
If you are interested in supporting GitHub-flavored alerts directly into MyST format, I just made a [Sphinx extension that does this automatically](https://kdeldycke.github.io/click-extra/sphinx.html#github-alerts).

It detects GitHub alerts in `include` directive and regular source file, and translate them to colon-fenced MyST admonitions behind the scenes.

Tl;Dr:
1. [Add `click-extra[sphinx]` dependency](https://kdeldycke.github.io/click-extra/install.html#for-sphinx) to your Sphinx build
2. [Add `click_extra.sphinx` to the `extensions` parameter](https://kdeldycke.github.io/click-extra/sphinx.html#id1) of your `conf.py` file
3. Activate `colon_fence` in `myst_enable_extensions`
4. That's it! :)

**Comment by webknjaz** [comment]
> Why is the [MyST reference documentation advertise support for GitHub alerts](https://mystmd.org/guide/admonitions#admonition-titles), while `MyST-Parser` is not supporting it?

@chrisjsewell could you comment on this?

**Comment by webknjaz** [comment]
Oh.. So that mystmd thing seems to be a TypeScript flavor.

**Comment by webknjaz** [comment]
Looks like there's something called `myst-transforms`: https://github.com/jupyter-book/mystmd/commit/e37d0c1b043a490ec4ee0e3c8b3d90e9a39c439c#diff-fce34afca1e1fe96424304c5ee8422076f20db3ccce8b79c2f4e024f9eee87acR7 / https://github.com/jupyter-book/mystmd/pull/514 / https://github.com/jupyter-book/mystmd/pull/229. Implemented by @rowanc1 in 2023 over there in TypeScript but not here.

Seems like `MyST-Parser` is out-of-sync, then? @chrisjsewell in the light of this discovery, would you reconsider having a plugin right in this project rather than external?

**Comment by chrisjsewell** [comment]
@webknjaz as I have requested before, please open a PR at https://github.com/executablebooks/mdit-py-plugins
